File: killrvideo-commons/src/main/java/com/killrvideo/dse/utils/PageableQueryFactory.java
package com.killrvideo.dse.utils;

import com.datastax.oss.driver.api.core.ConsistencyLevel;
import com.datastax.oss.driver.api.core.CqlSession;
import com.datastax.oss.driver.api.core.cql.BoundStatement;
import com.datastax.oss.driver.api.core.cql.BoundStatementBuilder;
import com.datastax.oss.driver.api.core.cql.PreparedStatement;
import com.datastax.oss.driver.api.core.cql.Row;
import com.datastax.oss.protocol.internal.util.Bytes;
import com.killrvideo.dse.dto.ResultListPage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;

import static com.killrvideo.dse.utils.AsyncResultSetUtils.toResultListPage;
import static org.apache.commons.lang3.StringUtils.isNotBlank;

@Component
public class PageableQueryFactory {
    private final CqlSession session;

    public PageableQueryFactory(CqlSession session) {
        this.session = session;
    }

    public <ENTITY> PageableQuery<ENTITY> newPageableQuery(String query,
            ConsistencyLevel consistencyLevel,
            Function<Row, ENTITY> rowMapper
    ) {
        return new PageableQueryImpl<>(query, this.session, consistencyLevel, rowMapper);
    }

    private static class PageableQueryImpl<ENTITY> implements PageableQuery<ENTITY> {
        private static final Logger LOGGER = LoggerFactory.getLogger(com.killrvideo.dse.utils.PageableQuery.class);

        private final CqlSession session;
        private final PreparedStatement preparedStatement;
        private final ConsistencyLevel consistencyLevel;
        private final Function<Row, ENTITY> rowMapper;

        PageableQueryImpl(String query,
                      CqlSession session,
                      ConsistencyLevel consistencyLevel,
                      Function<Row, ENTITY> rowMapper
        ) {
            this.session = session;
            this.preparedStatement = session.prepare(query);
            this.consistencyLevel = consistencyLevel;
            this.rowMapper = rowMapper;
        }

        public CompletableFuture<ResultListPage<ENTITY>> queryNext(
                Optional<Integer> pageSize,
                Optional<String> pageState,
                Object... queryParams) {
            BoundStatement statement = buildStatement(queryParams, pageSize, pageState);
            CompletableFuture<ResultListPage<ENTITY>> result = executeAsync(statement);
            if (LOGGER.isDebugEnabled()) {
                LOGGER.debug("Executed query {} with parameters: {}", statement.getPreparedStatement().getQuery(), queryParams);
            }
            return result;
        }

        private BoundStatement buildStatement(
                Object[] queryParams,
                Optional<Integer> pageSize,
                Optional<String> pageState) {
            BoundStatementBuilder builder = preparedStatement.boundStatementBuilder(queryParams);
            if (pageState.isPresent() && isNotBlank(pageState.get())) {
                builder.setPagingState(Bytes.fromHexString(pageState.get()));
            }
            pageSize.ifPresent(builder::setPageSize);
            builder.setConsistencyLevel(consistencyLevel);
            return builder.build();
        }

        private CompletableFuture<ResultListPage<ENTITY>> executeAsync(BoundStatement statement) {
            return this.session.executeAsync(statement).toCompletableFuture()
                    .thenApply(rs -> toResultListPage(rs, rowMapper));
        }
    }
}

File: killrvideo-commons/src/main/kotlin/com/killrvideo/dse/utils/PageableQueryFactory.kt
package com.killrvideo.dse.utils

import com.datastax.oss.driver.api.core.ConsistencyLevel
import com.datastax.oss.driver.api.core.CqlSession
import com.datastax.oss.driver.api.core.cql.AsyncResultSet
import com.datastax.oss.driver.api.core.cql.BoundStatement
import com.datastax.oss.driver.api.core.cql.PreparedStatement
import com.datastax.oss.driver.api.core.cql.Row
import com.datastax.oss.protocol.internal.util.Bytes
import com.killrvideo.dse.dto.ResultListPage
import mu.KotlinLogging
import org.apache.commons.lang3.StringUtils.isNotBlank
import org.springframework.stereotype.Component
import java.util.concurrent.CompletableFuture
import java.util.function.Function

@Component
class PageableQueryFactory(private val session: CqlSession) {
    fun <ENTITY> newPageableQuery(
        query: String,
        consistencyLevel: ConsistencyLevel,
        rowMapper: Function<Row, ENTITY>
    ): PageableQuery<ENTITY> =
        PageableQueryImpl(query, session, consistencyLevel, rowMapper)

    private class PageableQueryImpl<ENTITY> constructor(
        query: String,
        private val session: CqlSession,
        private val consistencyLevel: ConsistencyLevel,
        private val rowMapper: Function<Row, ENTITY>
    ) : PageableQuery<ENTITY> {
        private val logger = KotlinLogging.logger { }
        private val preparedStatement: PreparedStatement = session.prepare(query)

        override fun queryNext(
            pageSize: Int?,
            pageState: String?,
            vararg queryParams: Any
        ): CompletableFuture<ResultListPage<ENTITY>> {
            val statement = buildStatement(pageSize, pageState, arrayOf(*queryParams))
            val result = executeAsync(statement)
            logger.debug {"Executed query ${statement.preparedStatement.query} with parameters: $queryParams"}
            return result
        }

        private fun buildStatement(
            pageSize: Int?,
            pageState: String?,
            queryParams: Array<Any>
        ): BoundStatement {
            val builder = preparedStatement.boundStatementBuilder(*queryParams)
            if (isNotBlank(pageState)) {
                builder.setPagingState(Bytes.fromHexString(pageState))
            }
            pageSize ?.let { builder.setPageSize(pageSize) }
            builder.setConsistencyLevel(consistencyLevel)
            return builder.build()
        }

        private fun executeAsync(statement: BoundStatement): CompletableFuture<ResultListPage<ENTITY>> =
            session.executeAsync(statement).toCompletableFuture()
                .thenApply { rs: AsyncResultSet -> AsyncResultSetUtils.toResultListPage(rs, rowMapper) }
    }
}
